head.go

```diff
@@ -195,8 +227,30 @@ func (h *Head) ReadWAL() error {
 
 	// Track number of samples that referenced a series we don't know about
 	// for error reporting.
-	var unknownRefs int
+	var unknownRefs uint64
+
+	// Start workers that each process samples for a partition of the series ID space.
+	// They are connected through a ring of channels which ensures that all sample batches
+	// read from the WAL are processed in order.
+	var (
+		n          = runtime.GOMAXPROCS(0)
+		firstInput = make(chan []RefSample, 300)
+		input      = firstInput
+	)
+	for i := 0; i < n; i++ {
+		output := make(chan []RefSample, 300)
+
+		go func(i int, input <-chan []RefSample, output chan<- []RefSample) {
+			unknown := h.processWALSamples(mint, uint64(i), uint64(n), input, output)
+			atomic.AddUint64(&unknownRefs, unknown)
+		}(i, input, output)
+
+		input = output
```

Comment:
Can we have a comment here or somewhere that this is being done to reuse the sample buffer? (Is it what it is for?)

Reply:
Yes, done.